### DISCUSSION TITLE
Interaction between configurationProvider and other settings

### DISCUSSION THREAD

**Discussion by mmarczell-graphisoft** [microsoft/vscode-cpptools#7445] [Q&A]
Hello,

When I set 

```
"configurationProvider": "ms-vscode.cmake-tools",
"forcedInclude": ["something.h"]
```

the forcedInclude setting is not respected. Is that a bug or intended behaviour? How can I overcome this?

The scenario is Emscripten development on a Mac, and currently there is no way to tell the development environment that it shouldn't display this error when including iostream:

cannot open source file "Availability.h" (dependency of "iostream")

I tried the approach suggested in https://github.com/microsoft/vscode-cpptools/issues/1083#issuecomment-405523530 but it didn't work.

**Comment by Colengms** [comment]
Hi @mmarczell-graphisoft .  It's currently by-design that most other fields of the configuration are not used when using a custom configuration provider.  One exception is the compilerPath field, which if set to a blank string can be used to disable use of the compiler provider by a custom configuration provider, which still using detected include paths, etc..  Otherwise, the other fields of the configuration are only used when the custom configuration provider does not provide a configuration for a file.  It's expected that a custom configuration provider should provide a complete configuration for the file.

Since you are using CMake Tools, could you onfigure your CMakeLists.txt such that the forcedInclude is specified there, such as using a -include (gcc/clang) or /FI arg (cl.exe) ?

I'm not sure I understand the issue you are trying to work around.  Is that error coming from the C/C++ Extension?  Is "Availability.h" indeed a dependency of iostream?  Is that file missing?   Is the C/C++ extension failing to fully retrieve system include paths from the compiler?

**Reply by mmarczell-graphisoft** [reply]
A dependency of iostream is iterator.h, which on my Mac system includes this:

```
#ifdef __APPLE__
#include <Availability.h>
```

During an Emscripten compile run, `__APPLE__` is not defined, but the extension is assuming it is, so the editor displays these errors:

> #include errors detected. Consider updating your compile_commands.json or includePath. Squiggles are disabled for this translation unit
> cannot open source file "Availability.h" (dependency of "iostream")

This is why I tried the workaround in the above linked comment.

**Reply by sean-mcmanus** [reply]
It sounds like you need to change your intelliSenseMode to something that doesn't start with macos, but you'll probably need to make sure the compilerPath set doesn't return `__APPLE__` when queried.

**Reply by mmarczell-graphisoft** [reply]
It might be something obvious that I'm overlooking, but how do I manually query the compiler for defined macros?

**Reply by sean-mcmanus** [reply]
You can use `/usr/bin/clang -E -dD -x c++ /dev/null` (our extension will do a similar query when you set the compilerPath, you can see the output in the C/C++ logs when you set C_Cpp.loggingLevel to "Debug").

**Reply by mmarczell-graphisoft** [reply]
The emscripten compiler that I'm using does not define `__APPLE__`. My intelliSenseMode is set to `linux-clang-x64`, but in the extension debug log it seems like even that is being overridden. When I save my `c_cpp_properties.json`, I have 3 instances of this line in the extension log:

```
sending compilation args for /Volumes/SSD/Perforce/MM-Mac-Emscripten/main.cpp
```

and after the third one, when it has listed the defines for the third time, it ends with

```
  define: unix=1
  define: __unix=1
  define: __unix__=1
  other: --clang
  other: --clang_version=130000
  stdver: c++17
  intelliSenseMode: macos-clang-x64
```

**Reply by sean-mcmanus** [reply]
Are you using a configurationProvider? In that case the IntelliSense mode may be incorrectly set by the configuration provider.

Does your compiler define `__linux__`? If not, is there some other define we can check for?

**Reply by mmarczell-graphisoft** [reply]
As seen in my original post, I'm using the CMake Tools configuration provider.

**Reply by sean-mcmanus** [reply]
Can you check either the "C/C++" or "C++ Diagnostics" logging to see if the configuration provider is sending an intelliSenseMode? It would be in a section titled "Custom configurations". Also, if you stop using the configurationProvider and just set your compilerPath does the intelliSenseMode still get changed to macos-clang-x64?

**Reply by mmarczell-graphisoft** [reply]
I have this in the log:

```
Custom configurations received:
  uri: file:///Volumes/SSD/Me/main.cpp
  config: {
  "defines": [],
  "standard": "gnu++17",
  "includePath": [],
  "compilerPath": "/volumes/ssd/git/emsdk/upstream/emscripten/em++",
  "compilerArgs": [
    "-g",
    "-gsource-map",
    "--source-map-base",
    "file:////Volumes/SSD/Me/wasmbuild/",
    "-std=gnu++17"
  ]
}
```

If I don't set the configuration provider, IntelliSense doesn't get changed to macos.

**Reply by sean-mcmanus** [reply]
I'm guessing there may be a bug in how we handle those compilerArgs, which could be causing the query of em++ to fail and then we'd fall back to some other compiler (possibly due to https://github.com/microsoft/vscode-cpptools/issues/7269 which is fixed in 1.4.0-insiders and/or an issue with --source-map-base which we don't recognize as being an arg that takes an additional paramater...I can check/confirm later). Which compiler is shown as being used in the C/C++: Diagnostics logging?

**Reply by mmarczell-graphisoft** [reply]
Only the em++ compiler that I specified is ever mentioned in all of the logs. (For comparison, if I specify a nonexistent one, the logs tell me there was a fallback:

```
Specified compiler was not found: doesn'twork
Attempting to get defaults from compiler found on the machine: '/usr/bin/clang'
```
This does not happen with em++, it is found and queried.)

**Reply by sean-mcmanus** [reply]
Ok, can you confirm whether the em++ compiler defines `__unix__` but not `__linux__`? If so, we could switch to checking for that.

**Reply by mmarczell-graphisoft** [reply]
That's correct

**Reply by sean-mcmanus** [reply]
Thanks for the info. I've created an issue at https://github.com/microsoft/vscode-cpptools/issues/7525 .

**Reply by sean-mcmanus** [reply]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/1.4.0-insiders2 -- i.e. let us know if you're still hitting an issue.
